firedrake/slate/slac/compiler.py

```diff
@@ -533,6 +607,31 @@ def parenthesize(arg, prec=None, parent=None):
     return "(%s)" % arg
 
 
+def gem_to_loopy(gem_expr):
+    """ Method encapsulating stage 2.
+    Converts the gem expression dag into imperoc first, and then further into loopy.
+    :return slate_loopy: loopy kernel for slate operations.
+    """
+    # Creation of return variables for outer loopy
+    shape = gem_expr[0].children[0].shape
```

Comment:
The name suggests that `gem_expr` is a gem expression, but I think it's actually a list of gem expressions, which you only treat the first entry of. So require that it's a singleton list by doing:

```
expr, = gem_expressions
```

This additionally seems to suggest that the expression has a very specific form (namely an object with some free indices but no shape?). Can the previous `slate_to_gem` stage always return a shaped thing that we then index here?

Reply:
We index the gem_expr just because we wrapped a list around the mapper in the slate2gem translation. Will remove, so we can just use gem_expr. 
The second suggestion is good, then it becomes more clear that the most the indices of the top level node in the gem expression needs to match the output variable.